Added docs for virtual networks

sascala opened the pull request.

sascala commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+    $ iptables -I FORWARD -m set --match-set overlays src -m set --match-set overlays dst -m set ! --match-set src,dst overlay-exceptions -j dcos-isolation
+
+The actual iptables rules that live on the `dcos-isolation` chain are just simple rules, and we recommend that you again use ipsets of type `hash:net` and refer to `src` sets and `dest` sets. The purpose of this is primarily organization.
```
@mhausenblas Could you tell me a little more about what this sentence means?

sascala commented on 1.8/usage/virtual-networks.md:
```diff
@@ -0,0 +1,50 @@
+
+# Example
+
+The following Marathon application definition specifies a virtual network named `dcos-1`.
```
I'd like to say more about this example. What other aspects of it would be useful to call out?

asridharan replied:
@sascala what exactly do you need from me? Sorry was marked as the Assignee for this but couldn't figure out what is required?

sascala replied:
@asridharan Sorry not to be clear. Would you mind taking a quick look at the docs to make sure they're accurate/not missing anything important? I also posted a couple of questions.

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+menu_order: 11.5 
+---
+
+DC/OS allows you to create, manage, and use virtual networks, supporting IP-per-container with isolation guarantees amongst subnets. Virtual networks allow both Mesos and Docker containers to communicate from within a single node and between nodes on a cluster, and for services to be created so that their traffic is isolated from other traffic coming from any other virtual networks or hosts in th ...
```
isolation guarantees among subnets depends on the CNI implementation you're using and/or firewall policies that Dan implements...

jdef replied:
s/so that/such that/ ?

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+
+DC/OS allows you to create, manage, and use virtual networks, supporting IP-per-container with isolation guarantees amongst subnets. Virtual networks allow both Mesos and Docker containers to communicate from within a single node and between nodes on a cluster, and for services to be created so that their traffic is isolated from other traffic coming from any other virtual networks or hosts in th ...
+
+Additionally, virtual networks enable you to provide each container in the system with a unique IP address (“IP-per-container”). This removes the need to worry about potentially overlapping ports in applications, or the need to use nonstandard ports for services  to avoid overlapping. IP-per-container functionality also allows you to generate any number of instances of a class of tasks and have t ...
```
nit: remove extra whitespace s/`services  to`/`services to`/

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+
+Additionally, virtual networks enable you to provide each container in the system with a unique IP address (“IP-per-container”). This removes the need to worry about potentially overlapping ports in applications, or the need to use nonstandard ports for services  to avoid overlapping. IP-per-container functionality also allows you to generate any number of instances of a class of tasks and have t ...
+
+The virtual network enables you to run applications that require intra-cluster connectivity, like Cassandra, HDFS, and Riak. Your application can treat the various containers on your network like an end host, making it easier to enforce network isolation. This way, you can run multiple instances of DC/OS services without needing to manage ports. You can also create multiple virtual networks to is ...
```
s/The/A/ ?

jdef replied:
what is an "end host" ?

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+
+![Overview of the DC/OS Virtual Networks architecture](/1.8/administration/virtual-networks/img/virtual-networks.png) 
+
+DC/OS virtual networks do not require an external IP address management (IPAM) solution (and indeed at the current point in time do not support external IPAMs), since the global state (IP allocation) is handled via the Mesos Master replicated log. The interaction between the components looks like this:
```
please break up this sentence a bit, it's a bit overloaded

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+
+- Both the Mesos master and the Mesos agents run DC/OS overlay modules that communicate directly.
+
+- For the Mesos containerizer, the CNI isolator is used, whereas for the Docker containerizer [DNI](https://docs.docker.com/engine/userguide/networking/dockernetworks/) is leveraged, shelling out to the Docker daemon.
```
feels like comma overload

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+
+- For the Mesos containerizer, the CNI isolator is used, whereas for the Docker containerizer [DNI](https://docs.docker.com/engine/userguide/networking/dockernetworks/) is leveraged, shelling out to the Docker daemon.
+
+- For intra-node IP discovery we use an overlay orchestrator called navstar. This operator-facing system component is responsible for programming the overlay backend using a library called lashup that implements a gossip protocol to disseminate and coordinate overlay routing information amongst all Mesos agents in the DC/OS cluster.
```
do we have public documentation for navstar and lashup? worth linking to here?

sascala replied:
Not yet AFAIK.

asridharan replied:
@sargun ^^. I think `navstar` and `lashup` definitely should be documented. They are pretty cool and central to get the DC/OS overlay to work.

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+
+- For intra-node IP discovery we use an overlay orchestrator called navstar. This operator-facing system component is responsible for programming the overlay backend using a library called lashup that implements a gossip protocol to disseminate and coordinate overlay routing information amongst all Mesos agents in the DC/OS cluster.
+
+**Note:** To use virtual networks in DC/OS you must use a recent Linux kernel (3.9 or above) as well as Docker in version 1.11 on the agent nodes.
```
s/in version/version/

jdef replied:
also, docker 1.11 is the minimum version right? i'm assuming that we're aiming to be compatible with versions after that?

asridharan replied:
Yup it should docker 1.11 or higher.

jdef commented on 1.8/administration/virtual-networks/index.md:
```diff
@@ -0,0 +1,30 @@
+- For intra-node IP discovery we use an overlay orchestrator called navstar. This operator-facing system component is responsible for programming the overlay backend using a library called lashup that implements a gossip protocol to disseminate and coordinate overlay routing information amongst all Mesos agents in the DC/OS cluster.
+
+**Note:** To use virtual networks in DC/OS you must use a recent Linux kernel (3.9 or above) as well as Docker in version 1.11 on the agent nodes.
+Also note that we use the terms “virtual network” and “overlay network” in the context of DC/OS somewhat interchangeably.
```
can we just pick one to use for our documentation? I like "virtual network"

jdef replied:
ugh ... I see that our yaml examples (in other docs) actually have fields that use the words "overlay". somewhat unfortunate

asridharan replied:
@jdef I would want to stick with overlay networks at this point. I think virtual network is a much broader term that will encompass CNI in general. I think about DC/OS overlay networks as a particular virtual network, where the provider is DC/OS. Similarly Calico would be another virtual network with, Calico as the provider. 

This ties into the Network config service we are thinking about in DC/OS 1.9 (primarily for CNI networks). So would go with overlay networks here, and a promise for supporting more types of virtual networks going forward?

sascala replied:
Sounds like I should change most instances of "virtual network" to "overlay network" and mention in the overview that the overlay network is one of several types of virtual networks you can use on DC/OS.

asridharan replied:
I agree. @jdef ??

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
````diff
@@ -0,0 +1,194 @@
+        prefix: 24
+```
+
+**Note:** To use virtual networks in DC/OS you MUST make sure to use a recent Linux kernel (3.9 or above) as well as Docker version 1.11 on the agent nodes.
````
see prior comment about 1.11 being the minimum version

asridharan replied:
Yup, Docker 1.11 would be the minimum version. We will be compatible with anything higher than 1.11.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+
+**Note:** To use virtual networks in DC/OS you MUST make sure to use a recent Linux kernel (3.9 or above) as well as Docker version 1.11 on the agent nodes.
+
+Each overlay network is identified by a canonical `name`. Containers launched on an overlay network will get an IP address from the `subnet` allocated to the overlay network. To remove the dependency on a global IPAM (IP address management system), the overlay `subnet` is further split into smaller subnets, with each subnet being allocated to an agent. The agents can then use a host-local IPAM to ...
```
we define IPAM in the `index.md` file. why do it again here?

asridharan replied:
These sections were a written independently. So I am guess a bit of duplication in the heat of meeting 1.8 EA deadlines :).

sascala replied:
Correct.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+**Note:** To use virtual networks in DC/OS you MUST make sure to use a recent Linux kernel (3.9 or above) as well as Docker version 1.11 on the agent nodes.
+
+Each overlay network is identified by a canonical `name`. Containers launched on an overlay network will get an IP address from the `subnet` allocated to the overlay network. To remove the dependency on a global IPAM (IP address management system), the overlay `subnet` is further split into smaller subnets, with each subnet being allocated to an agent. The agents can then use a host-local IPAM to ...
+agent.
```
nit: please be consistent with newlines.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+Each overlay network is identified by a canonical `name`. Containers launched on an overlay network will get an IP address from the `subnet` allocated to the overlay network. To remove the dependency on a global IPAM (IP address management system), the overlay `subnet` is further split into smaller subnets, with each subnet being allocated to an agent. The agents can then use a host-local IPAM to ...
+agent.
+
+In the configuration above, each virtual network is allocated a /17 subnetwork (in the “subnet” field), which is then divided in /24 subnetworks to be used in each host that will be part of the network (in the “prefix” field). This allocates 8 bits (32 total minus 24 allocated for “prefix”) to designate the endpoint (container) inside a host for a maximum of 255 endpoints per host. It also alloca ...
```
good explanation. probably would be even better to follow it up with some more examples. there's "bit" of CIDR math here. examples always help.

sascala replied:
Nice pun!

I'd love to provide more examples. I tried to generate some myself but realized I'm missing rather a lot of information.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+
+# Adding overlay networks during installation
+
+Currently (as of version 1.8), DC/OS virtual networks can only be added and configured at install time. You can override the default network or add additional virtual networks by modifying your `config.yaml` file:
```
can I re-run the installer later to update my networks? if not .. is that something that we should make absolutely clear?

asridharan replied:
We don't support upgrades currently since we write these networks in the overlay replicated log. I think I was trying to bring this out clearly in the `re-install` section.

sascala replied:
I added some clarification on that right up front.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+    - 10.10.0.119
+    - 10.10.0.118
+    resolvers:
+    - 8.8.4.4
```
should we add a comment here in the resolvers section:
```
# you probably don't want to use these values since they point to public DNS servers.
# instead use values that are more specific to your particular infrastructure.
```

asridharan replied:
Agreed.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+
+In the above example, we have defined three overlay networks. The overlay networks `dcos-1` and `dcos-2` basically retain the default overlay networks, and we have added another overlay network called `dcos-3` with subnet range `44.128.0.0/16`. When you create a network, you must give it a name and a subnet. That name is used to launch Marathon tasks and other Mesos frameworks tasks using this sp ...
+
+**Note:** If you specify the same prefix and network mask for an overlay network, it will only work on one agent node.
```
Can/Should we detect this condition and warn the user about it?

sascala replied:
I'm not sure. @asridharan ?

asridharan replied:
The prefix defines the number of Agents on which the overlay can run. This is already explained in the previous section (I think), so seems like this NOTE is actually redundant?

sascala replied:
Indeed. I'll clarify the previous page.

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+
+# Retrieving overlay network state
+
+Once the DC/OS installation is completed you can query the virtual network configuration using the `https://leader.mesos/overlay-master/state` endpoint from within the cluster. In the following, we show the resulting JSON. The `network` key at the bottom lists the current overlay configuration and the `agents` key is a list showing how overlays are split across the Mesos agents. The following sho ...
```
again, sad that `overlay` found a home in in our API, this time in the form of a URL endpoint

asridharan replied:
@jdef any suggestions for better naming here? Maybe we can change it before GA?

jdef commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,194 @@
+
+**Note:** The overlay replicated log is different from the [master's replicated log](http://mesos.apache.org/documentation/latest/replicated-log-internals/), which is stored at /var/lib/mesos/master/replicated_log. Removing the *overlay* replicated log will have no effect on the master's recovery semantics.
+
+## iptables
```
can't we generated an uninstall script that helps the user do this?

sascala replied:
My understanding was that, for the moment at least, this is something the user needs to do manually. @asridharan - is my understanding accurate?

asridharan replied:
@sascala, yes you are right. Currently, the expectation is that user needs to perform this step manually. @jdef your point though is valid. Let me raise a JIRA and try and see if this can made part of the uninstall scripts.

jdef commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+# iptables rules
+
+DC/OS uses [iptables](http://linux.die.net/man/8/iptables), a fairly high speed, built-in mechanism for filtering traffic in Linux systems, to set up virtual network isolation. We recommend configuring filtering by deploying a homogenous set of rules to all nodes in your infrastructure. In order to simplify this, we also recommend using the [ipset](http://ipset.netfilter.org/ipset.man.html) featu ...
```
comma, overload, again

jdef commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+DC/OS uses [iptables](http://linux.die.net/man/8/iptables), a fairly high speed, built-in mechanism for filtering traffic in Linux systems, to set up virtual network isolation. We recommend configuring filtering by deploying a homogenous set of rules to all nodes in your infrastructure. In order to simplify this, we also recommend using the [ipset](http://ipset.netfilter.org/ipset.man.html) featu ...
+
+Set up your own chain that bumps over from the `FORWARD` chain. You can do this by running the following command:
```
"bumps over from" -- can we think of a different way to say this?

sascala replied:
It looks like "jump" is actually the correct verb: http://www.iptables.info/en/iptables-targets-and-jumps.html

jdef commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+We recommend using the `REJECT` directive as opposed to the `DROP` directive as it makes troubleshooting easier. The default is to allow all.
+
+Use ipset to get onto that chain. Create a `hash:net` type ipset named `overlays` that has all of the overlay networks that you want to restrict traffic from, or to. Then insert the rule:
```
s/that chain/the isolation chain/ ?

jdef commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+    $ iptables -I FORWARD -m set --match-set overlays src -m set --match-set overlays dst -j dcos-isolation
+
+This rule says that if a given packet is from any of the overlays and is destined to any other overlay, send it to the `dcos-isolation` rule. It can be advantageous to avoid filtering traffic from an overlay to itself. In order to avoid this, we recommend adding an exception set of type `hash:net,net` and adding entries for networks that should not be filtered. Modify the rule to: 
```
"advantageous" .. how? should we provide examples/reasons here?

jdef replied:
maybe this should be worded completely differently? "In most environments the system should not prevent a virtual network's block outbound packets from re-entering the same virtual network."

sascala replied:
I agree that providing a reason would be good here. @asridharan or @mhausenblas , can you explain why an administrator would want to avoid doing this?

asridharan replied:
@sargun I think would be the right person. @sargun ^^

jdef commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+The actual iptables rules that live on the `dcos-isolation` chain are just simple rules, and we recommend that you again use ipsets of type `hash:net` and refer to `src` sets and `dest` sets. The purpose of this is primarily organization.
+
+**Note:** In future versions of DC/OS we may provide support to automatically create the overlay ipsets for you. These would be under well-known names based on the set of overlays the user is running. For this reason, we strongly recommend not creating any ipsets that begin with `dcos-` or `mesos-`. 
```
Can we more simply state that network names prefixed with `dcos` or `mesos` reserved and will not be compatible with future releases?

sascala replied:
Since the support is hypothetical, it wouldn't be strictly true. But I can rephrase.

jdef commented on 1.8/administration/virtual-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+# Example
+
+Let’s say we have created two overlay networks: `IT` and `HR`. We want HR apps to be able to talk to IT apps, but no IT app should ever be allowed to connect to an HR app. We also want to allow all IT apps to talk to amongst themselves and all HR apps to talk amongst themselves. IT only runs apps on port 80. Let’s say that we’ve assigned HR an overlay with the agent subnets carved from `192.168.0 ...
```
make this the second-to-last sentence "We also want to allow all IT apps to talk to amongst themselves and all HR apps to talk amongst themselves."

and find a different way to say "let's say" rather than use it twice in the same paragraph

jdef commented on 1.8/usage/virtual-networks.md:
```diff
@@ -0,0 +1,50 @@
+menu_order: 8.5 
+---
+
+DC/OS allows you to use virtual networks to allow both Mesos and Docker containers to communicate from within a single node and between nodes on a cluster, and for services to be created so that their traffic is isolated from other traffic coming from any other virtual networks or hosts in the cluster.
```
s/so that/such that/ ?

jdef commented on 1.8/usage/virtual-networks.md:
```diff
@@ -0,0 +1,50 @@
+
+Additionally, virtual networks enable you to provide each container in the system with a unique IP address (“IP-per-container”). This removes the need to worry about potentially overlapping ports in applications, or the need to use nonstandard ports for services  to avoid overlapping. IP-per-container functionality also allows you to generate any number of instances of a class of tasks and have t ...
+
+The virtual network enables you to run applications that require intra-cluster connectivity, like Cassandra, HDFS, and Riak. Your application can treat the various containers on your network like an end host, making it easier to enforce network isolation. This way, you can run multiple instances of DC/OS services without needing to manage ports. You can also create multiple virtual networks to is ...
```
remove comma after "This way"

jdef commented:
There's lots of extra commas in these docs. I commented on a few of them. They're pretty distracting to read.

asridharan commented on 1.8/administration/virtual-networks/ip-per-container.md:
```diff
@@ -0,0 +1,196 @@
+
+Each overlay network is identified by a canonical `name`. Containers launched on an overlay network will get an IP address from the `subnet` allocated to the overlay network. To remove the dependency on a global IPAM, the overlay `subnet` is further split into smaller subnets, with each subnet being allocated to an agent. The agents can then use a host-local IPAM to allocate IP addresses from the ...
+
+In the configuration above, each virtual network is allocated a /17 subnetwork (in the “subnet” field), which is then divided in /24 subnetworks to be used in each host that will be part of the network (in the “prefix” field). This allocates 8 bits (32 total minus 24 allocated for “prefix”) to designate the endpoint (container) inside a host for a maximum of 255 endpoints per host. It also alloca ...
```
s/in /24 subnetworks/into /24 subnetworks

sascala commented:
OK, I've made most of the changes you've both suggested. Thanks for reading so carefully.

If anyone has an example or explanation of CIDR math on `ip-per-container.md`, I'd love to provide either or both.

Apart from that, and any other comments you may have, I'm assuming this doc is on hold until the overlay vs. virtual network discussion wraps up 1.8 EAs, whichever comes first. :-)

jgarcia-mesosphere commented:
At the risk of being too verbose, Wikipedia does a fair job of explaining it: https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#CIDR_notation

sascala commented:
Thanks, @jgarcia-mesosphere

sascala commented:
@jdef and @asridharan I've changed "virtual network" to "overlay network" everywhere it's relevant and added a (hopefully) clarifying image to `ip-per-container.md`. If everything looks OK to you, I'll send this on to an editor.

jdef commented on 1.8/administration/overlay-networks/index.md:
```diff
@@ -0,0 +1,42 @@
+
+- The CNI isolator is used for the Mesos containerizer. [DNI](https://docs.docker.com/engine/userguide/networking/dockernetworks/) is used for the Docker containerizer, shelling out to the Docker daemon.
+
+- For intra-node IP discovery we use an overlay orchestrator called navstar. This operator-facing system component is responsible for programming the overlay backend using a library called lashup that implements a gossip protocol to disseminate and coordinate overlay routing information among all Mesos agents in the DC/OS cluster.
```
link for lashup docs?

sascala replied:
I don't think there are any yet. Unless you know of some I don't?

jdef replied:
https://github.com/dcos/lashup

jdef commented on 1.8/administration/overlay-networks/index.md:
```diff
@@ -0,0 +1,42 @@
+
+- For intra-node IP discovery we use an overlay orchestrator called navstar. This operator-facing system component is responsible for programming the overlay backend using a library called lashup that implements a gossip protocol to disseminate and coordinate overlay routing information among all Mesos agents in the DC/OS cluster.
+
+**Note:** To use overlay networks in DC/OS you must use a recent Linux kernel (3.9 or later) as well as Docker version 1.11 and later on the agent nodes.
```
s/and later/or later/

jdef commented on 1.8/administration/overlay-networks/ip-per-container.md:
```diff
@@ -0,0 +1,200 @@
+menu_order: 10
+---
+
+In Enterprise DC/OS, the overlay network feature is enabled by default. The default configuration provides two overlay networks `dcos-1` and `dcos-2`, whose YAML configuration is as follows:
```
s/OS,/OS/

jdef commented on 1.8/administration/overlay-networks/ip-per-container.md:
```diff
@@ -0,0 +1,200 @@
+
+**Note:** Use a recent Linux kernel (3.9 or later) and Docker version 1.11 or later on the agent nodes.
+
+Each overlay network is identified by a canonical `name`. Containers launched on an overlay network will get an IP address from the `subnet` allocated to the overlay network. To remove the dependency on a global IPAM, the overlay `subnet` is further split into smaller subnets. Each of *those* subnets is allocated to an agent. The agents can then use a host-local IPAM to allocate IP addresses from ...
```
s/determines the size of the subnet/determines the size of the XXX subnet/

should we specify something for XXX that better qualifies subnet in this context?

sascala replied:
I agree. I edited to differentiate between the "overlay subnet" and "container subnet," but I'm open to other/better ideas.

jdef commented on 1.8/administration/overlay-networks/ip-per-container.md:
```diff
@@ -0,0 +1,200 @@
+
+Each overlay network is identified by a canonical `name`. Containers launched on an overlay network will get an IP address from the `subnet` allocated to the overlay network. To remove the dependency on a global IPAM, the overlay `subnet` is further split into smaller subnets. Each of *those* subnets is allocated to an agent. The agents can then use a host-local IPAM to allocate IP addresses from ...
+
+In the configuration above, each overlay network is allocated a /17 subnetwork (in the “subnet” field), which is then divided into /24 subnetworks to be used in each host that will be part of the network (in the “prefix” field), as shown:
```
s/above,/above/

jdef replied:
s/, as shown/as shown/

jdef commented on 1.8/administration/overlay-networks/ip-per-container.md:
```diff
@@ -0,0 +1,200 @@
+
+![Overlay network address space](/1.8/administration/overlay-networks/img/overlay-network-address-space.png)
+
+These values can be configured to adapt to each installation’s needs. If you need a larger network, decrease the network mask prefix of the `subnet`. To create an overlay network for only one agent node, specify the same value for the `subnet` network mask and the `prefix`. The size of the container space is `2^(32-<prefix>)`.
```
Why would I want an overlay network for only one node? Should we provide a use case for this?

sascala replied:
I'm not sure! @asridharan is there a reason a user might do this, or would it just be a mistake?

jdef commented on 1.8/administration/overlay-networks/ip-per-container.md:
```diff
@@ -0,0 +1,200 @@
+**Note:** The overlay replicated log is different from the [master's replicated log](http://mesos.apache.org/documentation/latest/replicated-log-internals/), which is stored at /var/lib/mesos/master/replicated_log. Removing the *overlay* replicated log will have no effect on the master's recovery semantics.
+
+## iptables
+The overlay networks install IPMASQ rules in order to allow containers to talk outside the overlay network. When you delete or replace overlay networks, you must remove the rules associated with the previous overlay networks. To remove the IPMASQ rules associated with each overlay, remove the IPMASQ rule from the POSTROUTING change of the NAT table that corresponds to the overlay networks subnet. ...
```
do we provide "comment" tags (or something else identifiable) for the `iptables` rules so that they're easy to target with a script?

sascala replied:
I do not know. @asridharan ?

jdef commented on 1.8/administration/overlay-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+    $ iptables -I FORWARD -m set --match-set overlays src -m set --match-set overlays dst -j dcos-isolation
+
+This rule says that if a given packet is from any of the overlays and is destined to any other overlay, send it to the `dcos-isolation` rule. In most environments, the system should prevent an overlay network's outbound packets from re-entering the same overlay network. In order to avoid this, add an exception set of type `hash:net,net` and add entries for networks that should not be filtered. Mo ...
```
s/system should prevent/system should not prevent/ ?

sascala replied:
How about "in most environments, the system does not prevent an overlay network's outbound packets from reentering the same overlay network. In order to prevent this..."?

And why would I want to ensure this doesn't happen?

jdef commented on 1.8/administration/overlay-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+
+The actual iptables rules that live on the `dcos-isolation` chain are just simple rules, and we recommend that you again use ipsets of type `hash:net` and refer to `src` sets and `dest` sets. The purpose of this is primarily organization.
+
+**Note:** In future versions of DC/OS we may provide support to automatically create the overlay ipsets for you. Network names prefixed with `dcos-` and `mesos-` are therefore reserved and should not be used.
```
The instructions tell people to set up `dcos-isolation` and here we say to avoid `dcos-` and `mesos-` prefixes. We should not contradict ourselves.

jdef replied:
or .. nevermind because this restriction only applies to ipsets and not to iptables chains?

sascala replied:
I believe that's correct.

jdef commented on 1.8/administration/overlay-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+Let’s say we have created two overlay networks: `IT` and `HR`. We want HR apps to be able to talk to IT apps, but no IT app should ever be allowed to connect to an HR app. We also want to allow all IT apps to talk to amongst themselves and all HR apps to talk amongst themselves. IT only runs apps on port 80. If we have assigned HR an overlay with the agent subnets carved from `192.168.0.0/16`, an ...
+
+    $ iptables -N dcos-isolation
+    $ iptables -A dcos-isolation -j REJECT # Changes it to default drop
```
a/drop/reject/

jdef commented on 1.8/administration/overlay-networks/isolation.md:
```diff
@@ -0,0 +1,73 @@
+    $ iptables -I dcos-isolation -m set --match-set hr src -m set --match-set hr dst -j RETURN
+    $ ipset add simple_allowed 192.168.0.0./16,192.168.0.0./16
+    $ ipset add simple_allowed 10.250.0.0/16,10.250.0.0/16
+    $ ipset add complex_allowed 192.168.0.0/16,80,10.250.0.0/16 #this allows traffic
```
should we expand on the comment here?

sascala replied:
Yes... it looks like that's the rule that allows traffic from HR to IT on port 80?

jdef commented on 1.8/usage/overlay-networks.md:
```diff
@@ -0,0 +1,60 @@
+
+* Both Mesos and Docker containers can communicate from within a single node and between nodes on a cluster.
+
+* Services can be created such that their traffic is isolated from other traffic coming from any other virtual network or host in the cluster.
```
s/virtual/overlay/ ?

jdef commented on 1.8/usage/overlay-networks.md:
```diff
@@ -0,0 +1,60 @@
+
+**Note:** Isolation guarantees among subnets depend on your CNI implementation and/or your firewall policies.
+
+# Using Virtual Networks
```
Overlay?

jdef commented on 1.8/usage/overlay-networks.md:
```diff
@@ -0,0 +1,60 @@
+
+First, you or the data center operator needs to [configure the virtual networks](/docs/1.8/administration/virtual-networks/).
+
+Virtual networks are configured at install time. You or the data center operator will specify a canonical name for each network in the `config.yaml`. When your service needs to launch a container, refer to it by that canonical name.
```
Overlay?

jdef commented on 1.8/usage/overlay-networks.md:
```diff
@@ -0,0 +1,60 @@
+
+# Using Virtual Networks
+
+First, you or the data center operator needs to [configure the virtual networks](/docs/1.8/administration/virtual-networks/).
```
virtual appears twice here, should be overlay?

jdef commented on 1.8/usage/overlay-networks.md:
```diff
@@ -0,0 +1,60 @@
+
+# Example
+
+The following Marathon application definition specifies a virtual network named `dcos-1`.
```
something more like this?

"specifies a virtual network name `dcos-1` which identifies the target DC/OS overlay network of the same name"

jdef commented on 1.8/administration/overlay-networks/index.md:
```diff
@@ -0,0 +1,42 @@
+
+* Both Mesos and Docker containers can communicate from within a single node and between nodes on a cluster.
+
+* Services can be created such that their traffic is isolated from other traffic coming from any other virtual network or host in the cluster.
```
s/virtual/overlay/ ?

jdef commented:
thanks for the update @sascala, left some comments

sascala commented:
much obliged, @jdef !

emanic commented:
👍 LGTM

jdef commented on 1.8/usage/overlay-networks.md:
```diff
@@ -20,15 +20,15 @@ DC/OS enables virtual networking through the use of overlay networks. DC/OS over
 # Example
 
-The following Marathon application definition specifies a virtual network named `dcos-1`.
+The following Marathon application definition specifies an overlay network named `dcos-1`, which refers to the target DC/OS overlay network of the same name.
```
nit: s/an overlay/a virtual/ -- because the marathon API calls it a "virtual network name"

sascala commented:
Done. Thank you, @jdef
